Tests often fail on macOS and Windows

When the tests are more stable, macOS and Windows could also be added to the GHA CI build matrix.

## macOS examples

See: https://github.com/wborn/jupnp/actions/runs/4347826782/jobs/7596808482

```
[INFO] Running TestSuite
[ERROR] Tests run: 322, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 420.718 s <<< FAILURE! - in TestSuite
[ERROR] byeByeBeforeAlive(org.jupnp.test.local.LocalDeviceBindingAdvertisementTest)  Time elapsed: 2.203 s  <<< FAILURE!
java.lang.AssertionError: expected [true] but found [false]
	at org.jupnp.test.local.LocalDeviceBindingAdvertisementTest.byeByeBeforeAlive(LocalDeviceBindingAdvertisementTest.java:178)

[ERROR] waitForRefresh(org.jupnp.test.local.LocalDeviceBindingAdvertisementTest)  Time elapsed: 2.158 s  <<< FAILURE!
java.lang.AssertionError: expected [true] but found [false]
	at org.jupnp.test.local.LocalDeviceBindingAdvertisementTest.waitForRefresh(LocalDeviceBindingAdvertisementTest.java:112)

[ERROR] moderateMaxRate(example.localservice.EventProviderTest)  Time elapsed: 0.549 s  <<< FAILURE!
java.lang.AssertionError: expected [four] but found [three]
	at example.localservice.EventProviderTest.moderateMaxRate(EventProviderTest.java:331)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   EventProviderTest.moderateMaxRate:331 expected [four] but found [three]
[ERROR]   LocalDeviceBindingAdvertisementTest.byeByeBeforeAlive:178 expected [true] but found [false]
[ERROR]   LocalDeviceBindingAdvertisementTest.waitForRefresh:112 expected [true] but found [false]
[INFO] 
[ERROR] Tests run: 322, Failures: 3, Errors: 0, Skipped: 0
```


## Windows examples

See: https://github.com/wborn/jupnp/actions/runs/4347826782/jobs/7596808992

```
[INFO] Running TestSuite
[ERROR] Tests run: 322, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 343.676 s <<< FAILURE! - in TestSuite
[ERROR] testPoolSize1(org.jupnp.test.service.QueueingThreadPoolExecutorTest)  Time elapsed: 1.01 s  <<< FAILURE!
java.lang.AssertionError: expected [2] but found [1]
	at org.jupnp.test.service.QueueingThreadPoolExecutorTest.testPoolSize1(QueueingThreadPoolExecutorTest.java:191)

[ERROR] testPoolSize1000ThreadsFast(org.jupnp.test.service.QueueingThreadPoolExecutorTest)  Time elapsed: 1.113 s  <<< FAILURE!
java.lang.AssertionError: Number of executors runs must match expected [1000] but found [1001]
	at org.jupnp.test.service.QueueingThreadPoolExecutorTest.testPoolSize1000ThreadsFast(QueueingThreadPoolExecutorTest.java:239)

[ERROR] testPoolSize10FillPoolParallel(org.jupnp.test.service.QueueingThreadPoolExecutorTest)  Time elapsed: 2.015 s  <<< FAILURE!
java.lang.AssertionError: expected [10] but found [7]
	at org.jupnp.test.service.QueueingThreadPoolExecutorTest.testPoolSize10FillPoolParallel(QueueingThreadPoolExecutorTest.java:608)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   QueueingThreadPoolExecutorTest.testPoolSize1:191 expected [2] but found [1]
[ERROR]   QueueingThreadPoolExecutorTest.testPoolSize1000ThreadsFast:239 Number of executors runs must match expected [1000] but found [1001]
[ERROR]   QueueingThreadPoolExecutorTest.testPoolSize10FillPoolParallel:608 expected [10] but found [7]
[INFO] 
[ERROR] Tests run: 322, Failures: 3, Errors: 0, Skipped: 0
```

--- a/.gitattributes
+++ b/.gitattributes
@@ -1,0 +1,2 @@
+*.java text=auto
+*.xml text=auto

--- a/.github/workflows/ci-build.yml
+++ b/.github/workflows/ci-build.yml
@@ -15,11 +15,17 @@ jobs:
       fail-fast: false
       matrix:
         java: [ '11' ]
-        maven: [ '3.9.4' ]
-        os: [ 'ubuntu-22.04' ]
+        maven: [ '3.9.6' ]
+        os: [ 'macos-12', 'ubuntu-22.04', 'windows-2022' ]
     name: Build (Java ${{ matrix.java }}, ${{ matrix.os }})
     runs-on: ${{ matrix.os }}
     steps:
+      - name: Set up line endings
+        if: ${{ matrix.os == 'windows-2022' }}
+        run: |
+          git config --global core.autocrlf false
+          git config --global core.eol crlf
+
       - name: Checkout
         if: github.head_ref == ''
         uses: actions/checkout@v3

--- a/bundles/org.jupnp/src/test/java/example/localservice/EventProviderTest.java
+++ b/bundles/org.jupnp/src/test/java/example/localservice/EventProviderTest.java
@@ -307,24 +307,24 @@ class EventProviderTest extends EventSubscriptionTest {
 
         upnpService.getControlPoint().execute(callback);
 
-        Thread.sleep(200);
+        Thread.sleep(800);
 
         Object serviceImpl = service.getManager().getImplementation();
 
         Reflections.set(Reflections.getField(serviceImpl.getClass(), "moderatedMaxRateVar"), serviceImpl, "two");
         service.getManager().getPropertyChangeSupport().firePropertyChange("ModeratedMaxRateVar", null, null);
 
-        Thread.sleep(200);
+        Thread.sleep(800);
 
         Reflections.set(Reflections.getField(serviceImpl.getClass(), "moderatedMaxRateVar"), serviceImpl, "three");
         service.getManager().getPropertyChangeSupport().firePropertyChange("ModeratedMaxRateVar", null, null);
 
-        Thread.sleep(200);
+        Thread.sleep(800);
 
         Reflections.set(Reflections.getField(serviceImpl.getClass(), "moderatedMaxRateVar"), serviceImpl, "four");
         service.getManager().getPropertyChangeSupport().firePropertyChange("ModeratedMaxRateVar", null, null);
 
-        Thread.sleep(100);
+        Thread.sleep(400);
 
         assertEquals(2L, callback.getSubscription().getCurrentSequence().getValue()); // It's the NEXT sequence!
 

--- a/bundles/org.jupnp/src/test/java/example/localservice/SwitchPowerModerated.java
+++ b/bundles/org.jupnp/src/test/java/example/localservice/SwitchPowerModerated.java
@@ -21,7 +21,7 @@ import example.binarylight.SwitchPower;
 
 public class SwitchPowerModerated extends SwitchPower {
 
-    @UpnpStateVariable(eventMaximumRateMilliseconds = 500)
+    @UpnpStateVariable(eventMaximumRateMilliseconds = 2000)
     public String moderatedMaxRateVar = "one";
 
     @UpnpStateVariable(eventMinimumDelta = 3)

--- a/bundles/org.jupnp/src/test/java/org/jupnp/local/LocalDeviceBindingAdvertisementTest.java
+++ b/bundles/org.jupnp/src/test/java/org/jupnp/local/LocalDeviceBindingAdvertisementTest.java
@@ -17,6 +17,10 @@ package org.jupnp.local;
 
 import static org.junit.jupiter.api.Assertions.*;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
 import org.junit.jupiter.api.Test;
 import org.jupnp.binding.LocalServiceBinder;
 import org.jupnp.binding.annotations.AnnotationLocalServiceBinder;
@@ -35,6 +39,7 @@ import org.jupnp.mock.MockUpnpServiceConfiguration;
 import org.jupnp.model.DiscoveryOptions;
 import org.jupnp.model.Namespace;
 import org.jupnp.model.ServerClientTokens;
+import org.jupnp.model.message.OutgoingDatagramMessage;
 import org.jupnp.model.message.UpnpMessage;
 import org.jupnp.model.message.header.UpnpHeader;
 import org.jupnp.model.meta.DeviceDetails;
@@ -59,7 +64,7 @@ class LocalDeviceBindingAdvertisementTest {
 
         upnpService.getRegistry().addDevice(binaryLight);
 
-        Thread.sleep(2000);
+        Thread.sleep(5000);
 
         assertEquals(12, upnpService.getRouter().getOutgoingDatagramMessages().size());
         for (UpnpMessage msg : upnpService.getRouter().getOutgoingDatagramMessages()) {
@@ -95,14 +100,17 @@ class LocalDeviceBindingAdvertisementTest {
         upnpService.getRegistry().addDevice(ld);
         assertEquals(1, upnpService.getRegistry().getLocalDevices().size());
 
-        Thread.sleep(2000);
+        Thread.sleep(5000);
 
         assertEquals(1, upnpService.getRegistry().getLocalDevices().size());
 
+        List<OutgoingDatagramMessage> outgoingDatagramMessages = new ArrayList<>(
+                upnpService.getRouter().getOutgoingDatagramMessages());
+
         // 30 from addDevice()
         // 30 from regular refresh
-        assertTrue(upnpService.getRouter().getOutgoingDatagramMessages().size() >= 60);
-        for (UpnpMessage msg : upnpService.getRouter().getOutgoingDatagramMessages()) {
+        assertTrue(outgoingDatagramMessages.size() >= 60);
+        for (UpnpMessage msg : outgoingDatagramMessages) {
             assertAliveMsgBasics(upnpService.getConfiguration().getNamespace(), msg, ld, 1);
         }
 
@@ -110,9 +118,14 @@ class LocalDeviceBindingAdvertisementTest {
 
         upnpService.shutdown();
 
+        // Ignore ALIVE messages send during shutdown
+        outgoingDatagramMessages = upnpService.getRouter().getOutgoingDatagramMessages().stream().filter(
+                msg -> msg.getHeaders().getFirstHeader(UpnpHeader.Type.NTS).getValue() == NotificationSubtype.BYEBYE)
+                .collect(Collectors.toList());
+
         // Check correct byebye
-        assertTrue(upnpService.getRouter().getOutgoingDatagramMessages().size() >= 30);
-        for (UpnpMessage msg : upnpService.getRouter().getOutgoingDatagramMessages()) {
+        assertTrue(outgoingDatagramMessages.size() >= 30);
+        for (UpnpMessage msg : outgoingDatagramMessages) {
             assertByeByeMsgBasics(upnpService.getConfiguration().getNamespace(), msg, ld, 1);
         }
     }
@@ -157,7 +170,7 @@ class LocalDeviceBindingAdvertisementTest {
 
         upnpService.getRegistry().addDevice(ld, new DiscoveryOptions(true, true));
 
-        Thread.sleep(2000);
+        Thread.sleep(5000);
 
         assertTrue(upnpService.getRouter().getOutgoingDatagramMessages().size() >= 60);
         // 30 BYEBYE

--- a/bundles/org.jupnp/src/test/java/org/jupnp/service/QueueingThreadPoolExecutorTest.java
+++ b/bundles/org.jupnp/src/test/java/org/jupnp/service/QueueingThreadPoolExecutorTest.java
@@ -17,6 +17,8 @@ package org.jupnp.service;
 
 import static org.junit.jupiter.api.Assertions.*;
 
+import java.time.Duration;
+import java.time.Instant;
 import java.util.Random;
 import java.util.StringTokenizer;
 import java.util.concurrent.SynchronousQueue;
@@ -184,25 +186,20 @@ class QueueingThreadPoolExecutorTest {
         pool.execute(createRunnable100ms());
         // queue thread must be active
         assertTrue(isQueueThreadActive(poolName));
-        // after 1+x sec all should be executed
-        Thread.sleep(1000);
-        assertEquals(2, pool.getCompletedTaskCount());
-        // at the end queue thread is shutdown, wait for additional 3 sec
-        Thread.sleep(3000);
-        assertFalse(isQueueThreadActive(poolName));
+        // all should be executed
+        waitForAssert(() -> assertEquals(2, pool.getCompletedTaskCount()));
+        // at the end queue thread is shutdown
+        waitForAssert(() -> assertFalse(isQueueThreadActive(poolName)));
 
         pool.execute(createRunnable100ms());
         pool.execute(createRunnable100ms());
         assertTrue(isQueueThreadActive(poolName));
-        // after 1+x sec all should be executed
-        Thread.sleep(1000);
-        assertEquals(4, pool.getCompletedTaskCount());
+        // all should be executed
+        waitForAssert(() -> assertEquals(4, pool.getCompletedTaskCount()));
 
-        // needs to wait CORE_POOL_TIMEOUT + 2sec-queue-thread + x until all
-        // threads are down again
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 3000);
-        assertFalse(areThreadsFromPoolRunning(poolName));
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     @Test
@@ -211,15 +208,13 @@ class QueueingThreadPoolExecutorTest {
         ThreadPoolExecutor pool = QueueingThreadPoolExecutor.createInstance(poolName, 2);
         pool.execute(createRunnableFast());
         pool.execute(createRunnableFast());
-        Thread.sleep(1000);
-        assertEquals(2, pool.getCompletedTaskCount());
+        waitForAssert(() -> assertEquals(2, pool.getCompletedTaskCount()));
 
         // no queue thread
 
-        // needs to wait CORE_POOL_TIMEOUT + x until all threads are down again
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 1000);
-        assertFalse(areThreadsFromPoolRunning(poolName));
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     @Test
@@ -232,14 +227,15 @@ class QueueingThreadPoolExecutorTest {
         }
         // no queue thread
         assertFalse(isQueueThreadActive(poolName));
-        Thread.sleep(1000);
-        assertEquals(1000, pool.getCompletedTaskCount(), "Completed tasks must match");
-        assertEquals(1000, AbstractRunnable.getRuns(), "Number of executors runs must match");
 
-        // needs to wait CORE_POOL_TIMEOUT + x until all threads are down again
+        waitForAssert(() -> {
+            assertEquals(1000, pool.getCompletedTaskCount(), "Completed tasks must match");
+            assertEquals(1000, AbstractRunnable.getRuns(), "Number of executors runs must match");
+        });
+
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 1000);
-        assertFalse(areThreadsFromPoolRunning(poolName));
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     @Test
@@ -252,14 +248,11 @@ class QueueingThreadPoolExecutorTest {
         pool.execute(createRunnableHeavyLoad1s());
 
         assertTrue(isQueueThreadActive(poolName));
-        Thread.sleep(5000);
-        assertEquals(4, pool.getCompletedTaskCount());
+        waitForAssert(() -> assertEquals(4, pool.getCompletedTaskCount()));
 
-        // needs to wait CORE_POOL_TIMEOUT + 2sec-queue-thread + x until all
-        // threads are down again
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 3000);
-        assertFalse(areThreadsFromPoolRunning(poolName));
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     @Test
@@ -311,11 +304,9 @@ class QueueingThreadPoolExecutorTest {
         // 0 are yet executed
         assertEquals(0, pool.getCompletedTaskCount());
 
-        // needs to wait CORE_POOL_TIMEOUT + 2sec-queue-thread + x until all
-        // threads are down again
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 3000);
-        assertFalse(areThreadsFromPoolRunning(poolName), "No threads must remain");
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     /**
@@ -349,11 +340,9 @@ class QueueingThreadPoolExecutorTest {
         // 0 are executed
         assertEquals(0, pool.getCompletedTaskCount());
 
-        // needs to wait CORE_POOL_TIMEOUT + 2sec-queue-thread + x until all
-        // threads are down again
-        // pool yet shutdown here
-        Thread.sleep(CORE_POOL_TIMEOUT + 3000);
-        assertFalse(areThreadsFromPoolRunning(poolName), "No threads must remain");
+        pool.shutdown();
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     /**
@@ -428,16 +417,12 @@ class QueueingThreadPoolExecutorTest {
             Thread.sleep(50); // chance for thread switch
             tEnd = System.currentTimeMillis();
         }
-        // chance to finalize
-        Thread.sleep(1000);
         // all should be executed
-        assertEquals(200, pool.getCompletedTaskCount());
+        waitForAssert(() -> assertEquals(200, pool.getCompletedTaskCount()));
 
-        // needs to wait CORE_POOL_TIMEOUT + 2sec-queue-thread + x until all
-        // threads are down again
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 3000);
-        assertFalse(areThreadsFromPoolRunning(poolName), "No threads must remain");
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     /**
@@ -461,11 +446,9 @@ class QueueingThreadPoolExecutorTest {
         // wait until all jobs have been processed
         Thread.sleep(2 * 1000 + 1000);
 
-        // needs to wait CORE_POOL_TIMEOUT + 2sec-queue-thread + x until all
-        // threads are down again
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 1000);
-        assertFalse(areThreadsFromPoolRunning(poolName), "No threads must remain");
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
 
         // all 5 jobs have to be executed
         assertEquals(5, AbstractRunnable.getRuns());
@@ -487,12 +470,10 @@ class QueueingThreadPoolExecutorTest {
         // queue thread must be active
         assertTrue(isQueueThreadActive(poolName));
         long queueThreadId1 = getThread(poolName + "-queue").getId();
-        // after 10+x sec all should be executed
-        Thread.sleep(12000);
-        assertEquals(100, pool.getCompletedTaskCount());
+        // all should be executed
+        waitForAssert(() -> assertEquals(100, pool.getCompletedTaskCount()));
         // at the end queue thread is shutdown, wait for additional 3 sec
-        Thread.sleep(3000);
-        assertFalse(isQueueThreadActive(poolName));
+        waitForAssert(() -> assertFalse(isQueueThreadActive(poolName)));
 
         // now put again requests into queue. Queue thread should be created
         // again with different thread id
@@ -501,22 +482,17 @@ class QueueingThreadPoolExecutorTest {
         }
         assertTrue(isQueueThreadActive(poolName));
         long queueThreadId2 = getThread(poolName + "-queue").getId();
-        // as queue thread has been created again, has to have different thread
-        // id
+        // as queue thread has been created again, has to have different thread id
         assertNotEquals(queueThreadId1, queueThreadId2);
 
-        // after 10+x sec all should be executed
-        Thread.sleep(12000);
-        assertEquals(200, pool.getCompletedTaskCount());
+        // all should be executed
+        waitForAssert(() -> assertEquals(200, pool.getCompletedTaskCount()));
         // at the end queue thread is shutdown, wait for additional 3 sec
-        Thread.sleep(3000);
-        assertFalse(isQueueThreadActive(poolName));
+        waitForAssert(() -> assertFalse(isQueueThreadActive(poolName)));
 
-        // needs to wait CORE_POOL_TIMEOUT + x until all
-        // threads are down again
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 1000);
-        assertFalse(areThreadsFromPoolRunning(poolName), "No threads must remain");
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     /**
@@ -539,9 +515,7 @@ class QueueingThreadPoolExecutorTest {
         for (Thread fillThread : fillThreads) {
             fillThread.start();
         }
-        Thread.sleep(2000); // wait until filled
-
-        assertEquals(10, pool.getActiveCount());
+        waitForAssert(() -> assertEquals(10, pool.getActiveCount()));
         // queue thread must be active
         assertTrue(isQueueThreadActive(poolName));
 
@@ -552,16 +526,13 @@ class QueueingThreadPoolExecutorTest {
         final int safetyTime = N * 10; // millis
         Thread.sleep(((N * millisPerRunnable) / poolSize) + safetyTime);
 
-        assertEquals(pool.getCompletedTaskCount(), N);
-        // at the end queue thread is shutdown, wait for additional 3 sec
-        Thread.sleep(3000);
-        assertFalse(isQueueThreadActive(poolName));
+        waitForAssert(() -> assertEquals(pool.getCompletedTaskCount(), N));
+        // at the end queue thread is shutdown
+        waitForAssert(() -> assertFalse(isQueueThreadActive(poolName)));
 
-        // needs to wait CORE_POOL_TIMEOUT + x until all
-        // threads are down again
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 1000);
-        assertFalse(areThreadsFromPoolRunning(poolName), "No threads must remain");
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     /**
@@ -608,11 +579,10 @@ class QueueingThreadPoolExecutorTest {
         for (Thread fillThread : fillThreads) {
             fillThread.start();
         }
-        Thread.sleep(1000); // wait until filled
 
-        assertEquals(5, pool.getActiveCount(), "All threads should be busy");
+        waitForAssert(() -> assertEquals(5, pool.getActiveCount(), "All threads should be busy"));
         // queue thread must be active
-        assertTrue(isQueueThreadActive(poolName));
+        waitForAssert(() -> assertTrue(isQueueThreadActive(poolName)));
 
         // wait until processed
         while (pool.getCompletedTaskCount() < 100) {
@@ -624,11 +594,9 @@ class QueueingThreadPoolExecutorTest {
         // check runs too
         assertEquals(100, AbstractRunnable.getRuns(), "Number of executors runs must match");
 
-        // needs to wait CORE_POOL_TIMEOUT + 2sec-queue-thread + x until all
-        // threads are down again
         pool.shutdown();
-        Thread.sleep(CORE_POOL_TIMEOUT + 3000);
-        assertFalse(areThreadsFromPoolRunning(poolName), "No threads must remain");
+        // after shutdown all threads are down again
+        waitForAssert(() -> assertFalse(areThreadsFromPoolRunning(poolName)));
     }
 
     // helper methods
@@ -732,6 +700,23 @@ class QueueingThreadPoolExecutorTest {
             }
         }
         return true;
+    }
+
+    private static void waitForAssert(Runnable runnable) {
+        final Instant start = Instant.now();
+        while (Duration.between(start, Instant.now()).toSeconds() < 20) {
+            try {
+                runnable.run();
+                return;
+            } catch (Error | NullPointerException e1) {
+                try {
+                    Thread.sleep(50);
+                } catch (InterruptedException e2) {
+                    throw new IllegalStateException("Interrupted while sleeping", e2);
+                }
+            }
+        }
+        runnable.run();
     }
 
     // Runnables for testing

--- a/bundles/org.jupnp/src/test/java/org/jupnp/ssdp/RegistryExpirationTest.java
+++ b/bundles/org.jupnp/src/test/java/org/jupnp/ssdp/RegistryExpirationTest.java
@@ -145,7 +145,7 @@ class RegistryExpirationTest {
 
         assertEquals(1, upnpService.getRegistry().getResources().size());
 
-        Thread.sleep(2000);
+        Thread.sleep(6000);
 
         assertTrue(testRunnable.wasExecuted);
 

